packages/openapi3/src/schema-emitter.ts

```diff
@@ -497,150 +472,18 @@ export class OpenAPI3SchemaEmitter extends TypeEmitter<
-      const additionalProps: Partial<OpenAPI3Schema> = mergeUnionWideConstraints
-        ? this.#applyConstraints(union, {})
-        : {};
-
-      if (mergeUnionWideConstraints && nullable) {
-        additionalProps.nullable = true;
-      }
-
-      if (Object.keys(additionalProps).length === 0) {
-        return new ObjectBuilder(schema);
-      } else {
-        if (
-          (schema instanceof Placeholder || "$ref" in schema) &&
-          !(type && shouldInline(program, type))
-        ) {
-          if (type && (type.kind === "Model" || type.kind === "Scalar")) {
-            return new ObjectBuilder({
-              type: "object",
-              allOf: B.array([schema]),
-              ...additionalProps,
-            });
-          } else {
-            return new ObjectBuilder({ allOf: B.array([schema]), ...additionalProps });
-          }
-        } else {
-          const merged = new ObjectBuilder<OpenAPI3Schema>(schema);
-          for (const [key, value] of Object.entries(additionalProps)) {
-            merged.set(key, value);
-          }
-          return merged;
-        }
-      }
-    };
-
-    const checkMerge = (schemaMembers: { schema: any; type: Type | null }[]): boolean => {
-      if (nullable) {
-        for (const m of schemaMembers) {
-          if (m.schema instanceof Placeholder || "$ref" in m.schema) {
-            return true;
-          }
-        }
-      }
-      return false;
-    };
-
-    if (schemaMembers.length === 0) {
-      if (nullable) {
-        // This union is equivalent to just `null` but OA3 has no way to specify
-        // null as a value, so we throw an error.
-        reportDiagnostic(program, { code: "union-null", target: union });
-        return new ObjectBuilder({});
-      } else {
-        // completely empty union can maybe only happen with bugs?
-        compilerAssert(false, "Attempting to emit an empty union");
-      }
-    }
-
-    if (schemaMembers.length === 1) {
-      return wrapWithObjectBuilder(schemaMembers[0], { mergeUnionWideConstraints: true });
-    }
-
-    const isMerge = checkMerge(schemaMembers);
-    const schema: OpenAPI3Schema = {
-      [ofType]: schemaMembers.map((m) =>
-        wrapWithObjectBuilder(m, { mergeUnionWideConstraints: isMerge }),
-      ),
-    };
-
-    if (!isMerge && nullable) {
-      schema.nullable = true;
-    }
-
-    if (discriminator) {
-      // the decorator validates that all the variants will be a model type
-      // with the discriminator field present.
-      schema.discriminator = { ...discriminator };
-      // Diagnostic already reported in compiler for unions
-      const discriminatedUnion = ignoreDiagnostics(getDiscriminatedUnion(union, discriminator));
-      if (discriminatedUnion.variants.size > 0) {
-        schema.discriminator.mapping = this.#getDiscriminatorMapping(discriminatedUnion);
-      }
-    }
-
-    return this.#applyConstraints(union, schema);
+  unionSchema(union: Union): ObjectBuilder<Schema> {
```

Comment:
I split out the discriminator logic into the base class (and was actually able to re-use that in one other place as well). The variant collection, merge check, and assertions are different between the 2 though due to how `null` is treated differently between the 2 specs. `wrapWithObjectBuilder` has only one difference right now (due to the `nullable` check that isn't relevant for 3.1) but as I mention elsewhere I think it's likely this function will change more.

If there's anything specifically you had in mind though please share!

Reply:
Got it. I feel like for general schema generation, we can split the function into several steps. Put the shared steps in base class and put changed steps in 3.0 or 3.1 class, so that the changes can be easily located and update.